bpo-34454: Fix issue with non-UTF8 separator strings

pganssle opened the pull request:
It is possible to pass a non-UTF-8 string as a separator in `datetime.isoformat`, but the current implementation starts by decoding to UTF-8, which will fail even for some valid strings.

In the special case of non-UTF-8 separators, we replace the separator character with `T` before encoding as UTF-8, so that encoding errors only occur on *invalid* ISO 8601 strings, and are handled as a standard `ValueError` (as would occur in the pure Python version). 

[bpo-34454](https://www.bugs.python.org/issue34454): Implementation of the fix without significant performance problems.

<!-- issue-number: [bpo-34454](https://www.bugs.python.org/issue34454) -->
https://bugs.python.org/issue34454
<!-- /issue-number -->

pganssle committed:
Fix issue with non-UTF8 separator strings

It is possible to pass a non-UTF-8 string as a separator in
datetime.isoformat, but the current implementation starts by decoding to
UTF-8, which will fail even for some valid strings.

In the special case of non-UTF-8 separators, we take a performance hit
by encoding the string as ASCII and replacing any invalid characters
with ?.

pganssle committed:
Fix non-UTF8 crash for (date|time)_fromisoformat

Previously this would end up dereferencing a NULL pointer if the
PyUnicode_AsUTF8AndSize call failed, this makes it so that the same
error as any other parsing error is raised.

pganssle committed:
Refactor non-UTF-8 sanitization

This increases performance for valid non-UTF-8 strings by avoiding
an error condition, and minimizes the impact on the rest of the
algorithm.

taleinat committed:
Add news entry for bpo-34454

taleinat committed:
Add tests for surrogate code points

Co-authored-by: Alexey Izbyshev <izbyshev@ispras.ru>
Co-authored-by: Paul Ganssle <paul@ganssle.io>

taleinat commented on Modules/_datetimemodule.c:
```diff
@@ -4839,6 +4852,37 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
+_sanitize_isoformat_str(PyObject* dtstr, unsigned char * needs_decref) {
+    Py_ssize_t len = PyUnicode_GET_LENGTH(dtstr);
+    *needs_decref = 0;
+    if (len < 10 || !Py_UNICODE_IS_SURROGATE(PyUnicode_READ_CHAR(dtstr, 10))) {
```
This should be `len < 11` or `len < 10`.

taleinat commented on Misc/NEWS.d/next/Library/2018-08-22-21-59-08.bpo-34454.z7uG4b.rst:
```diff
@@ -0,0 +1,4 @@
+Fix the .fromisoformat() methods of datetime types crashing when given
+unicode with non-UTF-8-encodable code points.  Specifically,
+datetime.fromisoformat() now accepts surrogate unicode code points used as
+the separator.
```
We should mention @izbyshev in this NEWS entry.

taleinat commented on Modules/_datetimemodule.c:
```diff
@@ -4839,6 +4852,41 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
 
+
+static PyObject *
+_sanitize_isoformat_str(PyObject *dtstr, unsigned char *needs_decref) {
```
IMO `needs_decref` should be an `int`, not `unsigned char`.

pganssle replied:
Makes no difference to me. Ideally it would be `bool` but I guess that's not a thing in C?

taleinat commented on Modules/_datetimemodule.c:
```diff
@@ -4839,6 +4852,41 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
+    // the separator; to allow datetime_fromisoformat to make the simplifying
+    // assumption that all valid strings can be encoded in UTF-8, this function
+    // replaces any surrogate character separators with `T`.
+    Py_ssize_t len = PyUnicode_GET_LENGTH(dtstr);
```
Using `PyUnicode_GET_LENGTH` requires having called `PyUnicode_READY` before. In this case (and below), just use `PyUnicode_GetLength()`.

taleinat commented on Modules/_datetimemodule.c:
```diff
@@ -4839,6 +4852,41 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
+    // replaces any surrogate character separators with `T`.
+    Py_ssize_t len = PyUnicode_GET_LENGTH(dtstr);
+    *needs_decref = 0;
+    if (len < 10 || !Py_UNICODE_IS_SURROGATE(PyUnicode_READ_CHAR(dtstr, 10))) {
```
Are you sure that only surrogates can cause failure to encode as UTF-8?

taleinat commented on Modules/_datetimemodule.c:
```diff
@@ -4883,20 +4939,32 @@ datetime_fromisoformat(PyObject* cls, PyObject *dtstr) {
+invalid_string_error:
+    PyErr_Format(PyExc_ValueError, "Invalid isoformat string: %R", dtstr);
+
+finally:
```
"finally" can be confusing since this doesn't happen upon success. I would just use "error".

taleinat reviewed (changes requested):
Looks good, just a few small details to amend.

taleinat commented on Modules/_datetimemodule.c:
```diff
@@ -4839,6 +4852,41 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
+    // replaces any surrogate character separators with `T`.
+    Py_ssize_t len = PyUnicode_GET_LENGTH(dtstr);
+    *needs_decref = 0;
+    if (len < 10 || !Py_UNICODE_IS_SURROGATE(PyUnicode_READ_CHAR(dtstr, 10))) {
```
This should be `len < 11` or `len <= 10`.

pganssle replied:
Fixed.

izbyshev commented on Modules/_datetimemodule.c:
```diff
@@ -4839,6 +4852,41 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
+        return NULL;
+    }
+
+    PyObject *str_out = PyUnicode_FromFormat("%UT%U", left, right);
```
I haven't checked it, but creating a copy of `dtstr` (with `PyUnicode_New`/ `PyUnicode_CopyCharacters` or similar) and using `PyUnicode_WriteChar` to replace the separator might be both faster and simpler than splitting the string and then joining it again.

pganssle replied:
OK, refactored. You are right, it is both simpler and faster.

izbyshev replied:
Great, thanks!

pganssle committed:
Refactor sanitize_isoformat_str

Rather than splitting the string at position 10 and re-joining it with
PyUnicode_Format, this copies the original unicode object and overwrites
the separator character.

Co-Authored-By: Alexey Izbyshev <izbyshev@ispras.ru>

taleinat commented on Modules/_datetimemodule.c:
```diff
@@ -4848,9 +4889,17 @@ datetime_fromisoformat(PyObject* cls, PyObject *dtstr) {
+    int needs_decref = 0;
+    dtstr = _sanitize_isoformat_str(dtstr, &needs_decref);
+
+    Py_ssize_t len = PyUnicode_GetLength(dtstr);
```
The `PyUnicode_GetLength()` here is now unnecessary.

izbyshev commented on Modules/_datetimemodule.c:
```diff
@@ -4848,9 +4889,17 @@ datetime_fromisoformat(PyObject* cls, PyObject *dtstr) {
 
-    Py_ssize_t len;
+    int needs_decref = 0;
+    dtstr = _sanitize_isoformat_str(dtstr, &needs_decref);
```
`dtstr` should be checked for NULL.

pganssle replied:
Good catch, C programming is hard. :(

Fixed now.

taleinat committed:
added mention of patch author in NEWS

serhiy-storchaka commented on Modules/_datetimemodule.c:
```diff
@@ -4839,6 +4852,33 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
 }
 
+static PyObject *
+_sanitize_isoformat_str(PyObject *dtstr, int *needs_decref) {
```
`{` should be on new line.

serhiy-storchaka commented on Modules/_datetimemodule.c:
```diff
@@ -4839,6 +4852,33 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
+    // the separator; to allow datetime_fromisoformat to make the simplifying
+    // assumption that all valid strings can be encoded in UTF-8, this function
+    // replaces any surrogate character separators with `T`.
+    Py_ssize_t len = PyUnicode_GetLength(dtstr);
```
`PyUnicode_GetLength()` can set an exception and return `-1`.

pganssle replied:
Hm, good call. For some reason I had the impression that `Py_ssize_t` was an unsigned integer.

serhiy-storchaka commented on Modules/_datetimemodule.c:
```diff
@@ -4839,6 +4852,33 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
+        return dtstr;
+    }
+
+    PyObject *str_out = PyUnicode_New(len, PyUnicode_MAX_CHAR_VALUE(dtstr));
```
`_PyUnicode_Copy()` could be used.

But I'm not sure that arbitrary character (including lone surrogates) should be accepted as a date-time separator. For example 2018-08-24016:10:00 looks pretty confusing.

pganssle replied:
When we developed `datetime.fromisoformat`, the contract of the function was that it should act as the inverse of `datetime.isoformat`, meaning that it should satisfy

```python
datetime.fromisoformat(dt.isoformat(*args, **kwargs)) == dt
```

for all values of `dt`, `args` and `kwargs`. Since `dt.isoformat(sep='\ud800')` and `dt.isoformat(sep='0')` are valid, we need to accept *any* character in the separator position in order to comply with the contract of the function.

It may be worth bringing up the appropriate contract in the `datetime-SIG` mailing list, but it was decided to use a very simple contract so that it's very simple to define what is and is not the correct behavior for this function (which could otherwise grow unwieldy).

serhiy-storchaka commented on Modules/_datetimemodule.c:
```diff
@@ -4883,20 +4934,32 @@ datetime_fromisoformat(PyObject* cls, PyObject *dtstr) {
     return dt;
+
+invalid_string_error:
+    PyErr_Format(PyExc_ValueError, "Invalid isoformat string: %R", dtstr);
```
This error message can contain not original string.

pganssle replied:
Oh good point. As part of the fixup I will change how this works so that `dtstr` is not re-assigned.
